zebra-state/src/service/chain_tip.rs

```diff
@@ -387,6 +391,38 @@ impl ChainTip for LatestChainTip {
         self.with_chain_tip_block(|block| block.transaction_hashes.clone())
             .unwrap_or_else(|| Arc::new([]))
     }
+
+    /// Returns when the state tip changes.
+    ///
+    /// Marks the state tip as seen.
+    #[instrument(skip(self))]
+    fn best_tip_changed(&mut self) -> BestTipChanged {
+        // The changed() future doesn't lock the value,
+        // so we don't need to use `with_chain_tip_block()` here.
+        //
+        // Clone the watch channel
+        let mut best_tip = self.receiver.clone();
+
+        // The cloned channel has the original change status,
+        // so we can mark the `self` channel as seen.
+        //
+        // Normally `changed()` does this, but we need the clone due to lifetimes.
+        self.mark_best_tip_seen();
```

Comment:
I'm not sure this would work, because `changed()` might mark a message as seen after waiting to be notified of a new message if the version is initially the same.

Reply:
That seems like the expected behaviour to me, but I'm not sure if we're both thinking of the same situation.

Can you give me a specific scenario where that behaviour is a problem?

Reply:
> Can you give me a specific scenario where that behaviour is a problem?

If `best_tip_changed` is called when the newest value in the channel has been marked seen, it would mark the next value received as seen in `best_tip` but not in `self`.  Though I think it would work fine as it's used in `get_block_template` anyways.

Reply:
I think this is the correct behaviour, because the next value gets marked as seen in `self` just before we query the state for it.

Reply:
(`best_tip` is dropped when the `changed()` future completes, so it doesn't matter when it is marked as changed, because it is never used again.)

Reply:
I've checked the docs and comments here and in the RPC implementation, and I think they cover what we've discussed here.

But if you'd like me to make changes to make it clearer, I'm happy to do that - feel free to ask questions about specific comments, or make suggestions.